exec/java-exec/src/main/java/org/apache/drill/exec/store/pcap/decoder/TcpSession.java

```diff
@@ -0,0 +1,334 @@
+    packetsFromReceiver = new ArrayList<>();
+
+    handshake = new TcpHandshake();
+    this.sessionID = sessionID;
+  }
+
+  /**
+   * This function adds a packet to the TCP session.
+   * @param p The Packet to be added to the session
+   */
+  public void addPacket(Packet p) {
+
+    // Only attempt to add TCP packets to session
+    if (!p.getPacketType().equalsIgnoreCase("TCP")) {
+      return;
+    }
+
+    // These variables should be consistent within a TCP session
+    if (packetCount == 0) {
+      srcIP = p.getSrc_ip();
+      dstIP = p.getDst_ip();
+
+      srcPort = p.getSrc_port();
+      dstPort = p.getDst_port();
+
+      srcMac = p.getEthernetSource();
+      dstMac = p.getEthernetDestination();
+      startTime = p.getTimestamp();
+    } else if (p.getSessionHash() != sessionID) {
+      logger.warn("Attempting to add session {} to incorrect TCP session.", sessionID);
+      return;
+    }
+
+    // Add packet to appropriate list and increment the data size counter
+    if (p.getSrc_ip().getHostAddress().equalsIgnoreCase(srcIP.getHostAddress())) {
+      packetsFromSender.add(p);
+      // Increment the data size counters
+      if (p.getData() != null) {
+        sentDataSize += p.getData().length;
+      }
+
+    } else {
+      packetsFromReceiver.add(p);
+      if (p.getData() != null) {
+        receivedDataSize += p.getData().length;
+      }
+    }
+
+    // Check flags if connection is not established
+    if (!handshake.isConnected()) {
+      if (p.getSynFlag() && p.getSrc_ip().getHostAddress().equalsIgnoreCase(srcIP.getHostAddress())) {
+        // This is part 1 of the TCP session handshake
+        // The host sends the first SYN packet
+        handshake.syn = true;
+        handshake.setSyn();
+        synTime = p.getTimestamp();
+      } else if (p.getSynFlag() && p.getAckFlag() && p.getSrc_ip().getHostAddress().equalsIgnoreCase(dstIP.getHostAddress())) {
+        // This condition represents the second part of the TCP Handshake,
+        // where the receiver sends a frame with the SYN/ACK flags set to the originator
+        handshake.synAck = true;
+        handshake.setAck();
+      } else if (p.getAckFlag() && p.getSrc_ip().getHostAddress().equalsIgnoreCase(srcIP.getHostAddress())) {
+        // Finally, this condition represents a successful opening of a TCP session, when the originator sends a frame with only the ACK flag set.
+        // At this point we finalize the session object and clear out the flags.
+        handshake.ack = true;
+        handshake.setAck();
+        ackTime = p.getTimestamp();
+        connectTime = ackTime - synTime;
+        //handshake.setConnected(sessionID);
+      }
+    } else {
+      /* Check for flags to close connection.  Closing a TCP session is more difficult than opening a session and there are
+      * a lot of ways that it can go wrong.  See https://accedian.com/enterprises/blog/close-tcp-sessions-diagnose-disconnections/ for references on
```

Comment:
```suggestion
      * a lot of ways that it can go wrong. See https://accedian.com/enterprises/blog/close-tcp-sessions-diagnose-disconnections/ for references on
```

Reply:
Fixed